doc/designs/black-codestyle.rst

```diff
@@ -0,0 +1,308 @@
+
+Adoption of |Black| will eliminate these issues and reduce |FreeIPA|'s
+`Python code style`_ to run ``make black``.
+
+Concerns
+========
+
+Auto-formatting |FreeIPA| with |Black| will change almost all Python files of
+FreIPA.
+
+* Mass-changes of code is going to make backports harder because there will
+  be a lot of merge conflicts. To reduce merge conflicts all active branches
+  have to be auto-formatted with |Black| at once.
+* Pull requests will also be affected by merge conflicts. Therefore we should
+  try to merge or close as many pull requests as possible before the code is
+  formatted.
+* Auto formatting is going to interfere with git history and make it harder
+  to find  issues. Git has an option to ignore revisions with
+  ``git blame --ignore-revs-file`` or ``blame.ignoreRevsFile`` option [1]_.
+  Once the formatting changes have landed in git, IPA can provide a
+  ``.gitignorerevs`` file that lists the formatting commits from all branches.
+* |Black| sometimes creates less readable code. It's possible to disable
+  reformatting with ``# fmt: off`` / ``# fmt: on``. This document proposes a
+  short list of :ref:`exceptions <black-disable-formatting>`.
+* Reformatting does never change the meaning of code. |Black| does not
+  perform modifications that affect the AST (abstract syntax tree). By
+  default the tool compares the AST before and after reformatting to ensure
+  correctness. This is also the reason why black cannot perform string
+  concatenation and sometimes creates line like
+  ``text = "multiline " "string"``.
+* In general |Black| spreads out long expression over multiple lines and
+  therefore can increases lines of code. The slightly larger line length
+  reduces lines of code in other cases. Reformatting increases the total
+  lines of code of all ``.py`` files by about 1% (765k before, 773k after).
+
+  When I first used |Black| in personal projects I was slightly annoyed by
+  this formatting style. Shortly later I started to appreciate the style.
+  It makes it easier to recognize begin and end of a list block, reduces
+  diffs when adding/removing an item, and encourages use of keyword arguments
+  in function calls.
+
+Open questions
+==============
+
+ipa-4-6 branch
+--------------
+
+The |FreeIPA| team still maintains ipa-4-6 branch for RHEL 7. The branch
+simplifies backports and fixes of IPA in RHEL.
+
+According to the `RHEL Life Cycle`_ document, RHEL 7 is slowly winding down.
+RHEL 7.8 was released March 31, 2020. RHEL 7.9 will be the last RHEL 7
+y-stream release. RHEL 7 has reached end of full support in August 2019 and
+is going to reach end of *Maintenance Support 1* in August 2020. During the
+*Maintenance Support 2* phase critical and import security fixes and select
+urgent bug fixes are provided until June 2024.
+
+I expect that the amount of backports to 4.6 will soon diminish to a few
+commits a year.
+
+Q: When can we start to limit backports to urgent CVEs only?
```

Comment:
According to the "RHEL Life Cycle" document above, "End of Maintenance Support or Maintenance Support 2" for RHEL7 happens on June 30, 2024. Maintenance Support 2 includes "Red Hat-defined Critical and Important impact Security Advisories (RHSAs) and selected (at Red Hat discretion) Urgent Priority Bug Fix Advisories (RHBAs)".

So there is no date that limits the changes to RHEL7 to urgent CVEs during the 10-year RHEL7 life cycle.